lnwallet/chainfee/filtermanager.go

```diff
@@ -0,0 +1,300 @@
+package chainfee
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/rpcclient"
+)
+
+const (
+	// fetchFilterInterval is the interval between successive fetches of
+	// our peers' feefilters.
+	fetchFilterInterval = time.Minute * 5
+
+	// medianWindow is the number of historical values that the
+	// filterManager will track.
+	//
+	// Assuming that each polling interval polls 8 outbound peers, the
+	// value of 48 gives us 6 polling intervals which is ~3 blocks worth of
+	// data.
+	medianWindow = 48
+
+	// minNumFilters is the minimum number of feefilters we need during a
+	// polling interval. If we have fewer than this, we won't consider the
+	// data.
+	minNumFilters = 6
+)
+
+var (
+	// errNoFilters is an error that's returned if fetchMedianFilter is
+	// called and there is no data available.
+	errNoFilters = fmt.Errorf("no data available")
+)
+
+// filterManager is responsible for determining an acceptable minimum fee to
+// use based on our peers' feefilter values.
+type filterManager struct {
+	// values stores the historical feefilter values we'll use to come up
+	// with an acceptable minimum feerate.
+	values    []SatPerKWeight
+	valuesMtx sync.RWMutex
+
+	fetchFunc func() ([]SatPerKWeight, error)
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+// newFilterManager constructs a filterManager with a callback that fetches the
+// set of peers' feefilters.
+func newFilterManager(cb func() ([]SatPerKWeight, error)) *filterManager {
+	f := &filterManager{
+		values: make([]SatPerKWeight, 0, medianWindow),
+		quit:   make(chan struct{}),
+	}
+
+	f.fetchFunc = cb
+
+	return f
+}
+
+// Start starts the filterManager.
+func (f *filterManager) Start() {
+	f.wg.Add(1)
+	go f.fetchPeerFilters()
+}
+
+// Stop stops the filterManager.
+func (f *filterManager) Stop() {
+	close(f.quit)
+	f.wg.Wait()
+}
+
+// fetchPeerFilters fetches our peers' feefilter values and adds them to our
+// historical values.
+func (f *filterManager) fetchPeerFilters() {
+	filterTicker := time.NewTicker(fetchFilterInterval)
+	defer filterTicker.Stop()
+
+	for {
+		select {
+		case <-filterTicker.C:
+			filters, err := f.fetchFunc()
+			if err != nil {
+				log.Errorf("Encountered err while fetching "+
+					"fee filters: %v", err)
+				return
+			}
+
+			f.updateWindow(filters)
```

Comment:
This I think is a step in the wrong direction, this now does an undifferentiated singular median calculation over both dimensions (peers * epochs). I'm still of the opinion that median selection makes sense per cohort but does not make sense wrt moving window because of lag.

Selecting the median per cohort makes sense because it means that half your peers will relay your transaction, but in a rising fee environment we lose that property if we are taking a median over the window.

Time jitter is fine. I don't actually think we need a window at all. We just need a singular "last remembered value" that if we get a new valid value we can override it, and if we don't we use the best value we have.